dali/pipeline/data/buffer_manager.h

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_DATA_BUFFER_MANAGER_H_
+#define DALI_PIPELINE_DATA_BUFFER_MANAGER_H_
+
+#include <map>
+#include <memory>
+#include <set>
+#include <utility>
+#include <vector>
+
+#include "dali/pipeline/operators/op_spec.h"
+#include "dali/pipeline/data/allocator_manager.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/pipeline/data/buffer.h"
+
+namespace dali {
+
+#define LINEAR_SEARCH
```

Comment:
Maybe it would be easier to just use multimap and then do lower_bound call to find first matching element?

Reply:
But the best match is not the smallest that fits the criteria, it is the one that is closest in size to what we want. So after finding the lower bound you would still need to traverse the multimap in case there are better candidates. And that is going to be slower than with vector.